src/eth.rs

```diff
@@ -1,75 +1,163 @@
 
-use tokio::sync::Mutex;
 
-#[derive(Serialize, Deserialize, Clone, Copy, Debug, PartialEq, Eq)]
-pub enum ProofBuddyMessageType {
-    SubmitProof,
-    Cancel,
-    InitiateChainlinkFinalization,
-    WithdrawEarnings,
+// Load the Banyan Contract ABI into Memory
+// Important: The ABI must be updated if the contract is updated
+lazy_static! {
+    static ref BANYAN_ABI_STR_REF: &'static str = include_str!("../abi/escrow.json");
 }
 
-pub struct VitalikProvider {
-    provider: Mutex<Provider<Http>>,
-    // TODO: eventually we will need to handle if the provider falls over halfway through submitting a transaction.
-    //my_pending_transactions: HashMap<TxnHash, PendingTransaction<'a>>,
-    contract: Mutex<Contract<Provider<Http>>>,
+/// EthClient - Everything needed to interact with Banyan's Ethereum Stack
+pub struct EthClient {
+    /// An Eth Provider. This is required to interact with the Ethereum Blockchain.
+    provider: Provider<Http>,
+    /// The chain ID of the network we're connected to. This is Required for signing transactions.
+    chain_id: u64,
+    /// An (optional) Eth Signer for singing transactions. This is required for interacting with payable functions.
+    signer: Option<SignerMiddleware<Provider<Http>, LocalWallet>>,
+    /// A Deployed Solidity Contract Address. This is required to interact with the Banyan Contract.
+    contract: Option<Contract<Provider<Http>>>,
+    /// A timeout for the Eth Client
+    timeout: Duration,
 }
 
-impl VitalikProvider {
-    pub fn new(url: String, contract_address: String) -> Result<Self> {
-        let provider = Provider::<Http>::try_from(url)?;
-        let provider2 = provider.clone();
-        let address = contract_address
-            .parse::<Address>()
-            .expect("could not parse contract address");
-        let abi: Abi = serde_json::from_str(
-            fs::read_to_string("contract_abi.json")
-                .expect("can't read file")
-                .as_str(),
-        )
-        .expect("couldn't load abi");
-
-        // TODO is this the right place to be sticking mutexes?
-        Ok(Self {
-            provider: Mutex::new(provider),
-            // my_pending_transactions: HashMap::new(),
-            //timeout: Duration::from_secs(timeout_seconds),
-            contract: Mutex::new(Contract::new(address, abi, provider2)),
-        })
+/// EthClientBuilder - A Default Builder for the EthClient
+impl Default for EthClient {
+    fn default() -> Self {
+        EthClientBuilder::default().build().unwrap()
+    }
+}
+
+#[derive(Clone, Debug, Copy, EthEvent)]
+struct NewOffer {
+    // #[indexed]
+    pub creator: Address,
+    // #[indexed]
+    pub executor: Address,
+    // #[indexed]
+    pub offerId: U256, // TODO: Snake case
+}
+
+/// The EthProvider is a wrapper around the ethers-rs Provider that handles all Ethereum interactions.
+impl EthClient {
+    /// Create a new default EthClientBuilder
+    pub fn builder() -> EthClientBuilder {
```

Comment:
why

Reply:
The builder pattern is cool. it let's you define different types of cusomization and initialization logic, and keeps your actual implementations clean. I found it made it easier to reason/type up all this biolerplate.

I can add a `new()` implementation if u want.

Reply:
Implemented

Reply:
deprected builder pattern, implemented new